crates/oxc_query/src/edges.rs

```diff
@@ -0,0 +1,2170 @@
+        adapter: &'a Adapter<'b>,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::parents(contexts, adapter)
+    }
+
+    pub(super) fn span<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::get_span(contexts)
+    }
+}
+
+pub(super) fn resolve_assignment_type_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "span" => assignment_type::span(contexts, resolve_info),
+        _ => {
+            unreachable!(
+                "attempted to resolve unexpected edge '{edge_name}' on type 'AssignmentType'"
+            )
+        }
+    }
+}
+
+mod assignment_type {
+    use trustfall::provider::{
+        ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo, VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+
+    pub(super) fn span<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::get_span(contexts)
+    }
+}
+
+pub(super) fn resolve_class_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "constructor" => class::constructor(contexts, resolve_info),
+        "entire_class_span" => class::entire_class_span(contexts, resolve_info),
+        "getter" => class::getter(contexts, resolve_info),
+        "method" => class::method(contexts, resolve_info),
+        "name_span" => class::name_span(contexts, resolve_info),
+        "property" => class::property(contexts, resolve_info),
+        "setter" => class::setter(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'Class'")
+        }
+    }
+}
+
+mod class {
+    use std::rc::Rc;
+
+    use oxc_ast::ast::{ClassElement, MethodDefinitionKind};
+    use trustfall::provider::{
+        resolve_neighbors_with, ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo,
+        VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+    use crate::vertex::{ClassMethodVertex, ClassPropertyVertex};
+
+    macro_rules! class_fn_edge_implem {
+        ($contexts:ident, $x:ident) => {
+            resolve_neighbors_with($contexts, |v| {
+                Box::new(v.as_class().unwrap_or_else(|| panic!("expected to have a class vertex, instead have: {v:#?}")).class.body.body.iter().filter_map(|class_el| {
```

Comment:
we using this unwrap_or_else a lot, I wonder if we can just change the .as_class() to instead be a result with an error that will generate this automatically?

Reply:
I think you're using the `EnumAsInner` crate's `as_class()` here, not Trustfall's, which makes the problem a bit harder.

Reply:
I am, yeah.